internal/file_unwrapper.go

```diff
@@ -0,0 +1,77 @@
+package internal
+
+import (
+	"archive/tar"
+	"github.com/pkg/errors"
+	"github.com/wal-g/tracelog"
+	"io"
+	"os"
+)
+
+type FileUnwrapperType int
+
+const (
+	DefaultBackupFileUnwrapper FileUnwrapperType = iota + 1
+	CatchupBackupFileUnwrapper
+)
+
+func NewFileUnwrapper(unwrapperType FileUnwrapperType, options *BackupFileOptions) IBackupFileUnwrapper {
+	switch unwrapperType {
+	case DefaultBackupFileUnwrapper:
+		return &DefaultFileUnwrapper{BackupFileUnwrapper{options}}
+	case CatchupBackupFileUnwrapper:
+		return &CatchupFileUnwrapper{BackupFileUnwrapper{options}}
+	default:
+		return &DefaultFileUnwrapper{BackupFileUnwrapper{options}}
+	}
+}
+
+type BackupFileOptions struct {
+	isIncremented bool
+	isPageFile    bool
+}
+
+type IBackupFileUnwrapper interface {
+	UnwrapNewFile(reader io.Reader, header *tar.Header, file *os.File) error
+	UnwrapExistingFile(reader io.Reader, header *tar.Header, file *os.File) error
```

Comment:
It's better to pass interfaces like Reader, Writer, ReaderWriter instead of a pointer to a file. It makes writing unit tests easier.

Reply:
Since we need file-specific methods later, like file.Sync(), file.WriteAt(), you propose to wrap all these methods in interface?

Reply:
Hm, I just mean that if you can extract some logic from using file, it's better to do it. Because it will be possible to test it. But if there is no logic without using files it will be useless.

Reply:
fixed in 4a91a430a19ebdcae18eebccf9a061de16a85156